### PR TITLE
Add generated-by annotation to CLI new-app

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://github.com/openshift/origin/pull/3987

### CODE REVIEW

**Comment by mfojtik** [comment]
@spadgett @csrwng  PTAL

**Comment by mfojtik** [comment]
[test]

**Comment by csrwng** [comment]
@mfojtik could we not just add the annotation once to the list of objects that new-app produces? (we can also distinguish between new-app and new-build if done in the cli calling code)

File: pkg/generate/app/app.go [hunk 1]
```diff
@@ -576,3 +581,12 @@ func LabelsFromSpec(spec []string) (map[string]string, []string, error) {
...
 	return labels, remove, nil
 }
+
+// addGeneratedByAnnotation adds openshift.io/generatedby to objects generated by
+// the new-app
+func addGeneratedByAnnotation(meta *kapi.ObjectMeta) {
+	if len(meta.Annotations) == 0 {
+		meta.Annotations = map[string]string{}
+	}
+	meta.Annotations["openshift.io/generatedby"] = "OpenShiftNewApp"
```
Comment (kargakis) [inline pkg/generate/app/app.go]: Make these constants

Comment (smarterclayton) [inline pkg/generate/app/app.go]: generated-by

Comment (spadgett) [inline pkg/generate/app/app.go]: Web console currently uses `openshift.io/generatedby`, so we'd want to change it there, too.

**Comment by danmcp** [comment]
@mfojtik Bump

**Comment by danmcp** [comment]
@mfojtik And now needs a rebase.

**Comment by mfojtik** [comment]
@csrwng @danmcp @smarterclayton rebased && fixed.

We do set annotations same way as we set the labels. We now set annotation for both new-build and new-app. There is some code duplication between setLabels and setAnnotations but I dunno if it is worth to make this more generic.

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
@csrwng @smarterclayton as a bonus I renamed "generatedby" to "generated-by" in both new-app and web console.

**Comment by mfojtik** [comment]
not sure what I broken in assets :-( @jwforres any ideas?

**Comment by liggitt** [comment]
Your Java console tar is out of date. Run clean-assets/install-assets/build-assets. There should be no diff in java/bindata.go

**Comment by mfojtik** [comment]
```console
π ~/go/src/github.com/openshift/origin ./hack/install-assets.sh 
[install-assets] npm install --unsafe-perm
[install-assets] node_modules/protractor/bin/webdriver-manager update
[ERROR] Command 'node_modules/protractor/bin/webdriver-manager update' failed with 8, logs:

module.js:340
    throw err;
          ^
Error: Cannot find module 'request'

```

File: pkg/generate/app/cmd/newapp.go [hunk 2]
```diff
@@ -29,6 +29,13 @@ import (
 	"github.com/openshift/origin/pkg/util/namer"
 )
 
+const GeneratedByNamespace = "openshift.io/generated-by"
+
+var (
+	GeneratedByNewApp   = map[string]string{GeneratedByNamespace: "OpenShiftNewApp"}
```
Comment (kargakis) [inline pkg/generate/app/cmd/newapp.go]: Exporting maps is awkward. Can you make all these consts and have `setAnnotations` construct the maps?

Comment (mfojtik) [inline pkg/generate/app/cmd/newapp.go]: I can, I wanted to have this on one place so you know how to change it :-) But you're right.

**Comment by mfojtik** [comment]
@kargakis thanks! fixed.

File: pkg/util/labels.go [hunk 3]
```diff
@@ -101,6 +101,81 @@ func AddObjectLabels(obj runtime.Object, labels labels.Set) error {
...
+		}
+
+		switch objType := obj.(type) {
+		case *deployapi.DeploymentConfig:
+			if err := addDeploymentConfigNestedAnnotations(objType, annotations); err != nil {
+				return fmt.Errorf("unable to add nested annotations to %s/%s: %v", accessor.Kind(), accessor.Name(), err)
+			}
+		}
+
+		if err := MergeInto(metaAnnotations, annotations, ErrorOnDifferentDstKeyValue); err != nil {
```
Comment (csrwng) [inline pkg/util/labels.go]: I would not throw an error on a different dstkeyvalue... if you are generating new objects based on existing objects generated by a different generator, you'd likely still want to set the new generated-by annotation.

File: pkg/util/labels.go [hunk 4]
```diff
@@ -101,6 +101,81 @@ func AddObjectLabels(obj runtime.Object, labels labels.Set) error {
...
+				}
+				if err := MergeInto(existing, annotations, OverwriteExistingDstKey); err != nil {
+					return err
+				}
+				m["annotations"] = mapToGeneric(existing)
+			}
+			return nil
+		}
+
+		// only attempt to set root labels if a root object called labels exists
```
Comment (csrwng) [inline pkg/util/labels.go]: s/labels/annotations/

File: pkg/util/labels.go [hunk 5]
```diff
@@ -101,6 +101,81 @@ func AddObjectLabels(obj runtime.Object, labels labels.Set) error {
...
+	if unstruct, ok := obj.(*runtime.Unstructured); ok && unstruct.Object != nil {
+		// the presence of "metadata" is sufficient for us to apply the rules for Kube-like
+		// objects.
+		// TODO: add swagger detection to allow this to happen more effectively
+		if obj, ok := unstruct.Object["metadata"]; ok {
+			if m, ok := obj.(map[string]interface{}); ok {
+
+				existing := make(map[string]string)
+				if l, ok := m["annotations"]; ok {
+					if found, ok := extractLabels(l); ok {
```
Comment (csrwng) [inline pkg/util/labels.go]: should probably rename the function since it's no longer used for labels only

**Comment by mfojtik** [comment]
@csrwng all comments addressed, thanks!

**Comment by mfojtik** [comment]
@csrwng @kargakis any other comments or this is good to merge?

**Comment by kargakis** [comment]
LGTM

**Comment by mfojtik** [comment]
[merge] !

**Pull request merged**